dot/node_test.go

```diff
@@ -0,0 +1,437 @@
+	}
+
+	cfg.SlotDuration = time.Second
+
+	if cfg.Telemetry == nil {
+		telemetryMock := NewMockClient(ctrl)
+		telemetryMock.EXPECT().SendMessage(gomock.Any()).AnyTimes()
+		cfg.Telemetry = telemetryMock
+	}
+
+	srvc, err := network.NewService(cfg)
+	require.NoError(t, err)
+
+	return srvc
+}
+
+func TestNodeInitialized(t *testing.T) {
+	cfg := NewTestConfig(t)
+	require.NotNil(t, cfg)
+
+	genFile := newTestGenesisRawFile(t, cfg)
+	require.NotNil(t, genFile)
+
+	cfg.Init.Genesis = genFile
+
+	nodeInstance := nodeBuilder{}
+	err := nodeInstance.initNode(cfg)
+	require.NoError(t, err)
+
+	type args struct {
+		basepath string
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "blank base path",
+			args: args{basepath: ""},
+			want: false,
+		},
+		{
+			name: "working example",
+			args: args{basepath: cfg.Global.BasePath},
+			want: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := NodeInitialized(tt.args.basepath); got != tt.want {
+				t.Errorf("NodeInitialized() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func initKeystore(cfg *Config) (*keystore.GlobalKeystore, error) {
+	ks := keystore.NewGlobalKeystore()
+	// load built-in test keys if specified by `cfg.Account.Key`
+	err := keystore.LoadKeystore(cfg.Account.Key, ks.Acco)
+	if err != nil {
+		logger.Errorf("failed to load account keystore: %s", err)
+		return nil, err
+	}
+
+	err = keystore.LoadKeystore(cfg.Account.Key, ks.Babe)
+	if err != nil {
+		logger.Errorf("failed to load BABE keystore: %s", err)
+		return nil, err
+	}
+
+	err = keystore.LoadKeystore(cfg.Account.Key, ks.Gran)
+	if err != nil {
+		logger.Errorf("failed to load grandpa keystore: %s", err)
+		return nil, err
+	}
+
+	// if authority node, should have at least 1 key in keystore
+	if cfg.Core.Roles == types.AuthorityRole && (ks.Babe.Size() == 0 || ks.Gran.Size() == 0) {
+		return nil, ErrNoKeysProvided
+	}
+
+	return ks, nil
+}
+
+func TestNode_StartStop(t *testing.T) {
+	serviceRegistryLogger := logger.New(log.AddContext("pkg", "services"))
+	type fields struct {
+		Name          string
+		Services      *services.ServiceRegistry
+		started       chan struct{}
+		metricsServer *metrics.Server
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		err    error
+	}{
+		{
+			name: "base case",
+			fields: fields{
+				Name:     "Node",
+				Services: services.NewServiceRegistry(serviceRegistryLogger),
+				started:  make(chan struct{}),
+			},
+			err: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			n := &Node{
+				Name:          tt.fields.Name,
+				Services:      tt.fields.Services,
+				started:       tt.fields.started,
+				metricsServer: tt.fields.metricsServer,
+			}
+
+			go func() {
+				<-n.started
+				n.Stop()
+			}()
+			err := n.Start()
+
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+			} else {
+				assert.NoError(t, err)
+			}
```

Comment:
Waiting and stopping in the goroutine might do all this after the test finishes; we want to stop the node before the (sub)test finishes.

Also I guess `n.started` is not written to if there was a start error right?

```suggestion
			err := n.Start()

			if tt.err != nil {
				assert.EqualError(t, err, tt.err.Error())
			} else {
				assert.NoError(t, err)
				<-n.started
				n.Stop()
			}
```

Reply:
When I change code as above, the test hangs and does not end.  So, I think there is an issue in the way I'm testing for start.